pcsx2/ps2/PGPU.cpp

```diff
@@ -0,0 +1,1110 @@
+	} else if (addr == PGPU_DAT_FIFO) { //reverse write - mind the direction set by reg f380
+		//PGpuDataReg = data;
+		ringBufPut(&pgifDatRbC, &data);
+	//	Console.WriteLn( "\n\r PGIF REVERSE !!! DATA write 0x%08X = 0x%08X  IF_CTRL= %08X   PGPU_STAT= %08X  CmdCnt 0x%X \n\r",  addr, data,  getUpdPgifCtrlReg(),  getUpdPgpuStatReg(), pgifCmdRbC.count);
+		drainPgpuDmaNrToIop();
+	} else if (addr == PGPU_STAT) {
+		PGpuStatReg = data; //Should all bits be writable?
+	} else if (addr == PGIF_CTRL) { //PGIF config
+		setPgifCtrlReg(data);
+		PGifCtrlRegOld = PGifCtrlReg; //data
+		fillFifoOnDrain(); //Now this checks the 0x8 bit of the PGIF_CTRL reg, so  it here too,
+		//so that it gets updated immediately once it is set.
+	} else if (addr == PGIF_1) {
+		pgif1reg = data; //GP0(E2h)
+	} else if (addr == PGIF_2) {
+		pgif2reg = data;
+	} else if (addr == PGIF_3) {
+		pgif3reg = data;
+	} else if (addr == PGIF_4) {
+		pgif4reg = data;
+	}
+	return;
+}
+
+u32 PGIFr(int addr) {
+	u32 data;
+	if (addr == PGPU_CMD_FIFO) {
+		cmdRingBufGet(&data);
+	} else 	if (addr == PGPU_DAT_FIFO) {
+		fillFifoOnDrain();
+		datRingBufGet(&data);
+	} else 	if (addr == PGPU_STAT) {
+		data = PGpuStatReg;
+	} else 	if (addr == PGIF_CTRL) { //PGIF config
+		data = getUpdPgifCtrlReg();
+	} else if (addr == PGIF_1) {
+		data = pgif1reg; //GP0(E2h)
+	} else if (addr == PGIF_2) {
+		data = pgif2reg;
+	} else if (addr == PGIF_3) {
+		data = pgif3reg;
+	} else if (addr == PGIF_4) {
+		data = pgif4reg;
+	}
+
+	#if REG_LOG == 1
+	if (addr != PGPU_DAT_FIFO)
+	if ((addr != PGIF_CTRL) || ((addr == PGIF_CTRL) && (getUpdPgifCtrlReg() != data)))
+	if (addr != PGPU_STAT)
+	Console.WriteLn( "PGIF read %08X = %08X  GPU_ST %08X  IF_STAT %08X IOPpc %08X EEpc= %08X ",  addr, data,  getUpdPgpuStatReg(), getUpdPgifCtrlReg(), psxRegs.pc, cpuRegs.pc);
+	#endif
+	return data;
+}
+
+void PGIFrQword(u32 addr, void* dat) {
+	u32 *data = (u32*)dat;
+
+	if (addr == PGPU_CMD_FIFO) { //shouldn't happen
+		Console.WriteLn( "PGIF QW CMD read =ERR!");
+	} else 	if (addr == PGPU_DAT_FIFO) {
+		fillFifoOnDrain();
+		datRingBufGet((u32*)((u32)data +0));
+		datRingBufGet((u32*)((u32)data +4));
+		datRingBufGet((u32*)((u32)data +8));
+		datRingBufGet((u32*)((u32)data +0xC));
+		//ringBufGet(&pgifDatRbC, (u32*)((u32)data +0));
+		//ringBufGet(&pgifDatRbC, (u32*)((u32)data +4));
+		//ringBufGet(&pgifDatRbC, (u32*)((u32)data +8));
+		//ringBufGet(&pgifDatRbC, (u32*)((u32)data +0xC));
+		fillFifoOnDrain();
+	} else {
+		Console.WriteLn( "PGIF QWord Read from add %08X  ERR - shouldnt happen!", addr);
+	}
+	#if REG_LOG == 1
+		if (addr != PGPU_DAT_FIFO)
+			Console.WriteLn( "PGIF QW read  0x%08X = %08X %08X %08X %08X ",  addr, *(u32*)(data +0), *(u32*)(data +1), *(u32*)(data +2), *(u32*)(data +3));
+	#endif
+	return;
+}
+
+void PGIFwQword(u32 addr, void* dat) {
+	u32 data = (u32)dat;
+	Console.WriteLn( "\n\r WARNING PGIF WRITE BY PS1DRV ! - NOT KNOWN TO EVER BE DONE! \n\r");
```

Comment:
`\n` instead of `\n\r` if the blanks lines are desired (I also note that it's usually `\r\n` on Windows).